Expose SDK event monitor

dnc40085 opened the pull request:
This PR exposes the SDK event monitor discussed in issue #870 
`wifi.sta.eventMonxxx` code was moved to `wifi_eventmon.c`
changed `wifi.sta.eventMonReg()`: callbacks can now be unregistered by leaving function field blank.
changed `wifi.sta.eventMonStop()`: all callbacks can now be unregistered by using `1` instead of `"unreg all"`
All backwards compatibility was preserved. The event monitor documentation can be found [here](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/docs/en/modules/wifi.md#wifieventmon-module).

Also included is the addition of the functions [`wifi.ap.deauth()`(click here for documentation)](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/docs/en/modules/wifi.md#wifiapdeauth) and [`wifi.sta.getrssi()`(click here for documentation)](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/docs/en/modules/wifi.md#wifistagetrssi)

pjsg commented:
Does the SDK event monitor callback execute at interrupt level or at base level? I would hope at base level.....

dnc40085 commented:
@pjsg Honestly, I have no idea how to find that out and the SDK documentation doesn't provide that much information, just the function wifi_set_event_handler_cb(), a very short description and some example code.

I thought rewording the commit would replace the original, oh well...

marcelstoer commented:
> I thought rewording the commit would replace the original, oh well...

If that's your intention then you need to squash/amend your commits and then force push them (`git push --force`). Force pushing will erase the comments based on code lines in the previous commit, though. The comments in this conversion will stay as they're not part of the repository.

dnc40085 commented:
I also discovered the undocumented function `wifi_softap_deauth` which goes perfectly with the event monitor so unauthorized stations can be deauth when they are connecting.

I have written an [implementation](https://github.com/dnc40085/nodemcu-firmware/blob/dev_wifi_ap_deauth/app/modules/wifi.c#L968) for the function, should I add it to this pull request?

jmattsson commented on app/include/rom.h:
```diff
@@ -131,6 +132,8 @@ int ets_printf(const char *format, ...)  __attribute__ ((format (printf, 1, 2)))
 
 int ets_sprintf(char *str, const char *format, ...)  __attribute__ ((format (printf, 2, 3)));
 
+int ets_vsprintf (char *d, const char *s, va_list ap);
```
This is not part of the boot ROM. It's provided by the SDK libs. This definition (and the old `ets_sprintf()` one above) should go into `sdk-overrides/include/ets_sys.h`.

dnc40085 replied:
Thank you for clarifying that, I wasn't sure where to put that.

jmattsson commented:
With the new `wifi.eventmon.register()`, does one still have to call `wifi.sta.eventMonStart()` ? What about if I'm only interested in monitoring AP events?

dnc40085 commented:
The functions `wifi.eventmon.register()` and `wifi.sta.eventMonReg()` are completely separate and can be used independently from one another and even disabled in `user_config.h`. 

The `wifi.sta.eventmon___()` functions use a timer to check `wifi_station_get_connect_status()` every 150ms and when the returned value changes the function executes a user registered callback.
The purpose of the function `wifi.eventMonStart()` is to arm the timer.

The function `wifi.eventmon.register()` registers the user's callback function so it can later be retrieved when the SDK calls the event handler `wifi_handle_event_cb`, which first checks for a user callback. Then, if found, it uses switch/case to determine which event was returned by the SDK then assembles the available information into a table and passes it to the user's callback function.

dnc40085 commented:
Event monitor now copies event data into a queue and posts a task(using new tasking interface) to process the queue

dnc40085 commented:
I'm guessing this failure is related to #1070

marcelstoer commented:
I restarted the build after I merged #1070 - all green now.

dnc40085 commented:
@marcelstoer Thank you, much appreciated.

@pjsg I was looking at the disassembled code for the the function `wifi_set_event_handler_cb` and I [found a call](https://gist.github.com/dnc40085/df31f74caadda00e7073#file-wifi_set_event_handler_cb-asm-L14) to the function `ets_timer_handler_isr`, so I'm inclined to say that the callback is executed at interrupt level rather than base level.
Since you posted your question I have changed the event monitor implementation to utilize the new tasking interface by @TerryE. 
The implementation I have written uses two separate functions to handle the event. The first function [`wifi_handle_event_cb`](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/app/modules/wifi_eventmon.c#L148)(registered with `wifi_set_event_handler_cb`) checks if there is a Lua callback registered for the corresponding event, if found, it then copies the event data into a queue(a linked list) then posts a low priority task to the second function, [`process_event_queue`](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/app/modules/wifi_eventmon.c#L175) which then processes the first event in the queue, forwards the data to the Lua callback, clears the processed event from the queue and lastly, if any events are remaining in the queue, it posts another task.

TerryE commented:
No need for it to go into RAM :)

devsaurus commented on app/include/user_modules.h:
```diff
@@ -47,6 +47,7 @@
 #define LUA_USE_MODULES_UART
 //#define LUA_USE_MODULES_UCG
 #define LUA_USE_MODULES_WIFI
+  #define LUA_USE_MODULES_WIFI_EVENTMON
```
Why the indent - did you intend to disable by default?

dnc40085 replied:
No, I indented it to suggest that it is a submodule of `wifi`

marcelstoer commented:
Not sure I fully understand this PR.

If I'm not mistaken the new module depends on the WiFi module, right? Then we have a issue with #386. What happens if you enable WiFi eventmon but not WiFi in `user_modules.h`? Also, we have a single `.md` file in the `docs` folder *per module*. This PR violates this.

I hope you can easily dispel my concerns.

dnc40085 commented:
It's not really a separate module, the event monitor code is just in a separate file but still effectively part of the `wifi` module and resides in `wifi.eventmon`. 

I separated the code from `wifi.c` following @TerryE suggestion that since `wifi.c` was getting a bit large, the event monitor code should be [split off into a wifi folder](https://github.com/nodemcu/nodemcu-firmware/issues/870#issuecomment-166703173). I tried to move the files into a new folder `app/modules/wifi` but I couldn't figure out how to get it to compile, since my makefile-fu is poor, so I left the files in `app/modules`.

As for the question about the entries in `user_modules.h`, if the `wifi` module is disabled and the event monitor is not, the event monitor code will be [excluded](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/app/modules/wifi_eventmon.c#L19) from the build, and the build will complete successfully devoid of all wifi and event monitor code. if it's an issue I could move it to `user_config.h`.

With the documentation, since the wifi event monitor code is part of the `wifi` module, should it stay in `wifi.md` or should it be split off into a separate file since `wifi.md` is getting big?

I have to remember to start writing more details in my PRs.

marcelstoer commented:
> It's not really a separate module, the event monitor code is just in a separate file but still effectively part of the wifi module

That's what I thought and that's why I was so surprised to see a separate entry in `user_modules.h`.
You know I maintain nodemcu-build.com so I'm definitely biased when I look at PRs. How do you suggest such a "half module" should be integrated there? Also, if you can enable WiFi but disable WiFi eventmon it'd be odd to look at the WiFi module docs and see functions there that won't work even though you WiFi enabled.

Still confused...

TerryE commented:
@dnc40085. Let me do a proper review today, before we consider committing this. I can sort out this file split stuff and any make magic needed.  One thing that I do notice is that you are breaking the convention  for namespaces in C: all public names should be prefixed with the the module name / namespace (e.g. `wifi_` in this case) or optionally `_wifi_` if "private" between wifi files. static (that is internal to file) entries can be named whatever you want but should avoid using another namespace prefix. So having static routines called lua_xxxx is a bad idea, but having extern ones is a no-no.

TerryE commented:
> Also, if you can enable WiFi but disable WiFi eventmon it'd be odd to look at the WiFi module docs and see functions there that won't work even though you WiFi enabled.

@marcelstoer  we already have this type of anomaly with SSL, SHA2, etc.  Sometimes documented functionality only works when the build is properly configured.  It's a compromise that we have made in the past when functionality takes a large footprint and the some builders might wish to disable if they are trying to configure  a build for 512kb Flash for example.

dnc40085 commented:
>you are breaking the convention for namespaces in C: all public names should be prefixed with the the module name / namespace ...having static routines called lua_xxxx is a bad idea, but having extern ones is a no-no.

@TerryE I didn't know about the namespace conventions, thanks for letting me know. As for extern, I previously thought that it was letting the compiler know that the function/variable was defined elsewhere, but after removing them from `wifi_common.h` I found that extern is only required for `wifi_eventmon_map[]`.
I added the functions `lua_table_add_xxx()` and `lua_cb_xxx()` in `wifi_common.h` partly because of the [DRY principal](https://en.wikipedia.org/wiki/Don't_repeat_yourself) but mainly to save a few bytes and when I was choosing the names I was only thinking in the scope of the `wifi` module not the whole project.

Would you like me to add a commit to fix these improper coding practices before you do your review?
 
>It's a compromise that we have made in the past when functionality takes a large footprint and the some builders might wish to disable if they are trying to configure a build for 512kb Flash for example. 

@TerryE For this reason I've been thinking of opening an issue suggesting that we make smart config (wifi.xxxsmart()) optional, considering that there must be others that don't require this feature and would much rather prefer having the 26k of extra flash to use for extra modules

Sorry for my bad coding practices. I'm always reading stuff to try to improve.

Also, t hanks for the suggestion on getting Pil 3rd ed., it's already clarified some stuff for me. So far I'm on page 40.
Any suggestions for a book on C?

TerryE commented:
> Would you like me to add a commit to fix these problems before you do your review?

What I was thinking  about going is just to do a PR into _your_  fork for you to look at and consider.  It is s entirely up to you if you want to accept this or do the rework yourself.  But it's sunny outside in the UK so I might just have some hours outside in the fresh air before I start :smile:

dnc40085 commented:
> It is's entirely up to you if you want to accept this or do the rework yourself

If you take the time to do a PR I'm not going to take credit for your work, but I will learn from it.

On my local branch I've already:
- Removed all the unnecessary externs
- Renamed `lua_table_add_xxx()` to `table_add_xxx()`
- Renamed `lua_cb_xxregister()` to `xxregister_lua_cb()`
- Renamed `event_task_id` to `wifi_event_monitor_task_id`
- Renamed `event_queue_xxx` to `wifi_event_queue_xxx`
- Renamed `process_event_queue` to `wifi_event_monitor_process_event_queue`
- Renamed `wifi_handle_event_cb` to `wifi_event_monitor_handle_event_cb`
- Renamed `event_monitor_reason_map` to `wifi_event_monitor_reason_map`
- Renamed `wifi_eventmon_map` to wifi_event_monitor_map

TerryE commented:
@dnc40085, sorry but I haven't had a chance to do this today.

TerryE commented:
I've made some comments on your fork.  As I said: all externs in the wifi namespace should be prefixed with ẁiif_`.  Als be vary when moving code out of line as this might in fact generate more code than you are saving.  It's always instructive looking at the generated code if the routine is only a few lines long.  If you move the code to a header as a static inline then gcc treats this as a macro so you get the source clarity without any runtime overhead.  Worth considering.

TerryE commented:
@dnc40085 I'll do a top to tail review of this this weekend and if OK then I'll merge it. Sorry for the delay.

dnc40085 commented:
No worries, I'm in no rush.

I was wondering... locally, I was reworking `wifi.sta.config` to be simpler(use a stack pointer) and I found a bug where station can't connect to an open AP and I wrote a fix, should I add another commit to this or should I submit a separate PR?

TerryE commented:
> Also, is [`LUA_USE_MODULES_WIFI_EVENTMON`](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/app/include/user_modules.h#L55) OK in `user_modules.h` or should I do away with it and just use the [entries](https://github.com/dnc40085/nodemcu-firmware/blob/dev_new_event_monitor/app/include/user_config.h#L98) I already have in `user_config.h`?

As far as the application programmer is concerned this is just a config option within `wifi`, so I would not have the extra `LUA_USE`  but stick with the enable option in the config file.

marcelstoer commented:
@dnc40085 We don't have an agreement as for #1146 but this PR is not (yet?) on that list, it may not make the cut. So, I feel it'd be helpful if you could create a dedicated PR for the `wifi.NULLMODE` fix. That one could then be merged rather sooner than later independently from this one. The relevant bug report #1034 has already been closed a while ago which is confusing.

TerryE commented:
Perhaps we should apply a 90:10 test "Is this a material step forward over the current code in master/dev?".  I am always worried that we contributors are all time limited and if we apply a 100% rule then we'll just end up never moving forward.  

On this basis I would merge this now, but not without at least one other contributor +1.

dnc40085 commented:
So what should I do? split it off?

TerryE commented:
@dnc40085 

> I took a look at the disassembled code for the the function `wifi_set_event_handler_cb` and I [found a call](https://gist.github.com/dnc40085/df31f74caadda00e7073#file-wifi_set_event_handler_cb-asm-L14) to the function `ets_timer_handler_isr`, so I'm inclined to say that the callback is executed at interrupt level rather than base level.

If you look gist then you will see that this s offset 0xCC which is in `ets_timer_init()` so I am inclined to differ. :smile: 

Nope IMO leave it as it is now now.  90:10 rules.

jmattsson commented:
This PR has been lingering for some time now even after all the updates. What are the arguments against merging this? Is it the "half-module" that Marcel brought up? If that's the only thing I'm in favour of merging now and improving later.

dnc40085 commented:
If anybody has any questions about the PR please feel free to ask. 
Suggestions for changes are also welcome.

dnc40085 commented:
@TerryE 
>If you look gist then you will see that this is offset by 0xCC which is in `ets_timer_init()` so I am inclined to differ. :smile:

You're probably right, my understanding of assembler is limited.

TerryE commented:
@dnc40085 are you happy for us to merge this?

dnc40085 commented:
I have no more changes to make to this PR, Aside from not being sure if I should remove the pseudo [future proofing section](https://github.com/nodemcu/nodemcu-firmware/pull/1018/files#diff-7396aef6e0a69a3cca8825fb77305aedR196) (If Espressif adds any new events, this section allows the user to assign a default callback to execute if an event specific callback is not defined).

Should I squash the commits before the merge?

jmattsson commented:
Yes please., might as well keep it neat and tidy!

dnc40085 commented:
The commits have been squashed.

jmattsson commented:
Thanks! Sorry about the delay to get this merged.

dnc40085 commented:
No problem. Thanks for merging this.
